core/remotes/docker/lister.go

```diff
@@ -0,0 +1,144 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package docker
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+
+	"github.com/containerd/errdefs"
+	"github.com/containerd/log"
+)
+
+var ErrObjectNotRequired = errors.New("object not required")
+
+type TagList struct {
+	Name string   `json:"name"`
+	Tags []string `json:"tags"`
+}
+
+type dockerLister struct {
+	dockerBase *dockerBase
+}
+
+func (r *dockerLister) List(ctx context.Context) ([]string, error) {
+	refSpec := r.dockerBase.refspec
+	base := r.dockerBase
```

Comment:
This implementation looks far from complete as the OCI distribution spec also defines pagination; https://github.com/opencontainers/distribution-spec/blob/v1.1.0/spec.md#listing-tags


I need to look up references, but I _think_ registries are allowed to limit the number of tags returned, in which case pagination may be needed to get all tags.

I know there's been some  repositories with literally millions of tags, which would be problematic to return in a single response, and responses  for such repositories may be limited on the registry side.

Reply:
Oh that is a good find. Thanks.

Reply:
I wonder though... We, might not want to actually return millions of tags... So we might even need some kind of user side "setting" to limit that?

Reply:
Reading through this, I'm not sure this would work nicely. Like, there is nothing in the response saying something like "we have more tags" or "last page". It just says you can play around with what we show you.

That means we will always have tags + 1 calls to list that uses the last tag as lasttag name to see if there are more tags. That could be rather inefficient. But it is how it is, if that's the case.